batch/batch/driver/main.py

```diff
@@ -1322,6 +1330,150 @@ async def monitor_system(app):
     monitor_instances(app)
 
 
+async def compact_agg_billing_project_users_table(db: Database):
+    @transaction(db)
+    async def compact(tx: Transaction):
+        target = await tx.execute_and_fetchone(
+            '''
+SELECT billing_project, `user`, resource_id
+FROM (
+  SELECT billing_project, `user`, resource_id, COUNT(*) AS n_tokens
+  FROM aggregated_billing_project_user_resources_v3
+  WHERE token != 0
+  GROUP BY billing_project, `user`, resource_id
+) AS t
+ORDER BY n_tokens DESC
+LIMIT 1;
+''',
+            query_name='find_agg_billing_project_user_resource_to_compact',
+        )
+
+        if target is None:
+            return
+
+        original_usage = await tx.execute_and_fetchone(
+            '''
+SELECT CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s
+FOR UPDATE;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        # overwriting the value for token = 0 here is intentional because we've computed the total usage above
+        await tx.execute_update(
+            '''
+INSERT INTO aggregated_billing_project_user_resources_v3 (billing_project, `user`, resource_id, token, `usage`)
+VALUES (%s, %s, %s, %s, %s)
+ON DUPLICATE KEY UPDATE `usage` = %s
+''',
+            (target['billing_project'], target['user'], target['resource_id'], 0, original_usage, original_usage),
+        )
+
+        await tx.just_execute(
+            '''
+DELETE FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s AND token != 0;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        new_usage = await tx.execute_and_fetchone(
+            '''
+SELECT billing_project, `user`, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s
+GROUP BY billing_project, `user`, resource_id;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        if new_usage != original_usage:
```

Comment:
are these dictionaries guaranteed to be the same? Should we instead check `new_usage['usage'] == original_usage['usage']`?